src/map.mo

```diff
@@ -0,0 +1,305 @@
+    /// assert(e.lookupOrPut("aaa", "b") == ?("b", 1));
+    /// assert(e.lookupOrPut("abc", "c") == ?("a", 0);
+    /// ```
+    /// Runtime: O(key_size) acesses to stable memeory.
+    public func lookupOrPut(key : Blob, value : Blob) : ?(Blob, Nat) {
+      let { leaves; nodes } = base.regions();
+
+      let ?leaf = base.put_(nodes, leaves, key) else return null;
+      let ret_value = if (leaf == base.leaf_count - 1) {
+        base.setValue(leaves, leaf, value);
+        value;
+      } else {
+        base.getValue(leaves, leaf);
+      };
+      ?(ret_value, Nat64.toNat(leaf));
+    };
+
+    /// Returns `?(index, value)` where `index` is the index of `key` in order it was added to enumeration and `value` is corresponding value to the `key`,
+    /// or `null` it `key` wasn't added.
+    ///
+    /// Example:
+    /// ```motoko
+    /// let e = StableTrie.Map({
+    ///   pointer_size = 2;
+    ///   aridity = 2;
+    ///   root_aridity = null;
+    ///   key_size = 2;
+    ///   value_size = 1;
+    /// });
+    /// assert(e.put("abc", "a") == ?0);
+    /// assert(e.put("aaa", "b") == ?1);
+    /// assert(e.lookup("abc") == ?("a", 0);
+    /// assert(e.lookup("aaa") == ?("b", 1));
+    /// assert(e.lookup("bbb") == null);
+    /// ```
+    /// Runtime: O(key_size) acesses to stable memeory.
+    public func lookup(key : Blob) : ?(Blob, Nat) {
+      base.lookup(key);
+    };
+
+    public func delete(key : Blob) : ?Blob {
+      let { leaves; nodes } = base.regions();
+      let bytes = Blob.toArray(key);
+
+      let idx = base.keyToRootIndex(bytes);
+      let child = base.getChild(nodes, 0, idx);
+      let (value, leaf_opt) = deleteRec(nodes, leaves, bytes, child, base.root_bitlength);
+      switch (leaf_opt) {
+        case (#single leaf) {
+          base.setChild(0, idx, leaf);
+        };
+        case (#empty) {
+          base.setChild(0, idx, 0);
+        };
+        case (_) {};
+      };
+      value;
+    };
+
+    type DeleteRes = {
+      #empty;
+      #multiple;
+      #single : Nat64;
+    };
+
+    public func singleLeaf(region : Base.Region, node : Nat64) : DeleteRes {
+      let blob = Region.loadBlob(region.region, base.getOffset(node, 0), base.node_size_);
+      let bytes = Blob.toArray(blob);
+
+      var lastNode : Nat64 = 0;
+      for (i in Iter.range(0, args.aridity - 1)) {
+        var x : Nat64 = 0;
+        for (i in Iter.revRange(i * args.pointer_size + args.pointer_size - 1, i * args.pointer_size)) {
+          x := x * 256 + Nat64.fromNat(Nat8.toNat(bytes[Int.abs(i)]));
+        };
+        if (x > 0) {
+          if (lastNode != 0) return #multiple;
+          lastNode := x;
+        };
+      };
+      if (lastNode == 0) return #empty;
+      if (lastNode & 1 == 0) return #multiple;
+      return #single lastNode;
+    };
+
+    func deleteRec(nodes : Base.Region, leaves : Base.Region, bytes : [Nat8], node : Nat64, pos : Nat16) : (?Blob, DeleteRes) {
+      if (node == 0) return (null, #empty);
```

Comment:
If we return `#empty` here then the code above will do one unnecessary Region access. It will over write 0 with 0. Can we arrange that differently? Maybe we need a new to return here instead of `#empty`?